File: python/features/steps/id_generation_steps.py
"""Behave steps for issue ID generation."""

from __future__ import annotations

import re

from behave import given, then, when

from taskulus.ids import (
    IssueIdentifierRequest,
    generate_issue_identifier,
    generate_many_identifiers,
    set_test_uuid_sequence,
)


@given('a project with project key "{project_key}"')
def given_project_key(context: object, project_key: str) -> None:
    context.id_prefix = project_key
    context.existing_ids = set()


@given('a project with an existing issue "{identifier}"')
def given_project_with_existing_issue(context: object, identifier: str) -> None:
    context.existing_ids = {identifier}
    context.id_prefix = identifier.split("-")[0]


@when("I generate an issue ID")
def when_generate_issue_id(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    existing = getattr(context, "existing_ids", set())
    request = IssueIdentifierRequest(
        title="Test title",
        existing_ids=existing,
        prefix=prefix,
    )
    context.generated_id = generate_issue_identifier(request).identifier


@when("I generate 100 issue IDs")
def when_generate_many_issue_ids(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    context.generated_ids = generate_many_identifiers("Test title", prefix, 100)


@given('the UUID generator always returns "{uuid_text}"')
def given_uuid_generator_returns(context: object, uuid_text: str) -> None:
    set_test_uuid_sequence([uuid_text] * 11)


@when("I attempt to generate an issue ID")
def when_attempt_generate_issue_id(context: object) -> None:
    prefix = getattr(context, "id_prefix", "tsk")
    existing = getattr(context, "existing_ids", set())
    request = IssueIdentifierRequest(
        title="Test title",
        existing_ids=existing,
        prefix=prefix,
    )
    try:
        context.generated_id = generate_issue_identifier(request).identifier
        context.id_generation_error = None
    except RuntimeError as error:
        context.generated_id = None
        context.id_generation_error = str(error)


@then('the ID should match the pattern "{pattern}"')
def then_id_matches_pattern(context: object, pattern: str) -> None:
    regex = re.compile(f"^{pattern}$")
    assert regex.match(context.generated_id)


@then("all 100 IDs should be unique")
def then_ids_are_unique(context: object) -> None:
    assert len(context.generated_ids) == 100


@then('the ID should not be "{forbidden}"')
def then_id_not_collision(context: object, forbidden: str) -> None:
    assert context.generated_id != forbidden


@then('ID generation should fail with "{message}"')
def then_id_generation_failed(context: object, message: str) -> None:
    assert context.id_generation_error == message

File: rust/features/steps/id_generation_steps.rs
use std::collections::HashSet;

use cucumber::{given, then, when};
use regex::Regex;

use taskulus::ids::{
    generate_issue_identifier, generate_many_identifiers, set_test_uuid_sequence,
    IssueIdentifierRequest,
};
use uuid::Uuid;

use crate::step_definitions::initialization_steps::TaskulusWorld;

#[given(expr = "a project with project key {string}")]
fn given_project_key(world: &mut TaskulusWorld, project_key: String) {
    world.id_prefix = Some(project_key);
    world.existing_ids = Some(HashSet::new());
}

#[given(expr = "a project with an existing issue {string}")]
fn given_project_existing_issue(world: &mut TaskulusWorld, identifier: String) {
    let mut existing = HashSet::new();
    existing.insert(identifier.clone());
    world.existing_ids = Some(existing);
    let prefix = identifier.split('-').next().unwrap_or("tsk");
    world.id_prefix = Some(prefix.to_string());
}

#[when("I generate an issue ID")]
fn when_generate_issue_id(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let existing = world.existing_ids.clone().unwrap_or_default();
    let request = IssueIdentifierRequest {
        title: "Test title".to_string(),
        existing_ids: existing,
        prefix,
    };
    let result = generate_issue_identifier(&request).expect("generate identifier");
    world.generated_id = Some(result.identifier);
}

#[when("I generate 100 issue IDs")]
fn when_generate_many_ids(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let ids = generate_many_identifiers("Test title", &prefix, 100).expect("generate ids");
    world.generated_ids = Some(ids);
}

#[given(expr = "the UUID generator always returns {string}")]
fn given_uuid_generator_returns(_world: &mut TaskulusWorld, uuid_text: String) {
    let parsed = Uuid::parse_str(&uuid_text).expect("parse uuid");
    set_test_uuid_sequence(Some(vec![parsed; 11]));
}

#[when("I attempt to generate an issue ID")]
fn when_attempt_generate_issue_id(world: &mut TaskulusWorld) {
    let prefix = world.id_prefix.clone().unwrap_or_else(|| "tsk".to_string());
    let existing = world.existing_ids.clone().unwrap_or_default();
    let request = IssueIdentifierRequest {
        title: "Test title".to_string(),
        existing_ids: existing,
        prefix,
    };
    match generate_issue_identifier(&request) {
        Ok(result) => {
            world.generated_id = Some(result.identifier);
            world.id_generation_error = None;
        }
        Err(error) => {
            world.generated_id = None;
            world.id_generation_error = Some(error.to_string());
        }
    }
}

#[then(expr = "the ID should match the pattern {string}")]
fn then_id_matches_pattern(world: &mut TaskulusWorld, pattern: String) {
    let identifier = world.generated_id.as_ref().expect("generated id");
    let regex = Regex::new(&format!("^{pattern}$")).expect("regex");
    assert!(regex.is_match(identifier));
}

#[then("all 100 IDs should be unique")]
fn then_ids_unique(world: &mut TaskulusWorld) {
    let ids = world.generated_ids.as_ref().expect("generated ids");
    assert_eq!(ids.len(), 100);
}

#[then(expr = "the ID should not be {string}")]
fn then_id_not_collision(world: &mut TaskulusWorld, forbidden: String) {
    let identifier = world.generated_id.as_ref().expect("generated id");
    assert_ne!(identifier, &forbidden);
}

#[then(expr = "ID generation should fail with {string}")]
fn then_id_generation_failed(world: &mut TaskulusWorld, message: String) {
    let error = world
        .id_generation_error
        .as_ref()
        .expect("id generation error");
    assert_eq!(error, &message);
}
